README.md
# IMPLICON - bisulfite amplicon data for imprinted loci

This is a processing guide for IMPLICON data, with detailed intructions of how to get from raw data to methylation consistancy plots at imprinted loci. 

Publication at [NAR](https://academic.oup.com/nar/article/48/16/e92/5867412): **IMPLICON: an ultra-deep sequencing method to uncover DNA methylation at imprinted regions**



Last update: 28/03/2021

### Table of Contents
* [Quick Start](#quick-start)
* [Detailed Processing Guide](#detailed-data-processing-guide)
  1. [UMI-handling](#step-i-umi-handling)
  2. [Adapter-/Quality Trimming](#step-ii-adapter-quality-trimming)
  3. [Genome Alignments](#step-iii-genome-alignments)
    - [Allele-specific alignments (optional)](#allele-specific-alignments)
  4. [UMI-aware deduplication](#step-iv-umi-aware-deduplication)
  5. [Allele-specific sorting (optional)](#step-v-allele-specific-sorting) 
  6. [Methylation extraction](#step-vi-methylation-extraction)
* [Read-level methylation consistency analysis](#read-level-methylation-consistency)
  1. [Filtering Implicon CpG context](#filtering-cpg-context-files)
  2. [Plotting read-level methylation consistency](#plotting-read-level-methylation-consistency)

## QUICK START

The following commands are designed to work with a hypothetical example paired-end IMPLICON dataset consisting of reads from a C57BL/6 mouse:

**Read 1:** `test_R1.fastq.gz`
**Read 2:** `test_R2.fastq.gz`

**Step I: UMI-handling**

```
trim_galore --paired --implicon *fastq.gz
```

**Step II: Adapter-/quality trimming**

```
trim_galore --paired *UMI*fastq.gz
```

**Step III: Genome alignments**

```
bismark --genome /Genomes/Mouse/GRCm38/ -1 test_8bp_UMI_R1_val_1.fq.gz -2 test_8bp_UMI_R2_val_2.fq.gz
```
\* for allele-specific alignments see [below](step-iii-genome-alignments)

**Step IV: UMI-aware deduplication**

```
deduplicate_bismark --barcode test_8bp_UMI_R1_val_1_bismark_bt2_pe.bam
```


**Step V: Allele-specific sorting**

\* for allele-specific sorting please see [below](step-v-allele-specific-sorting)


**Step VI: Methylation extraction**

```
bismark_methylation_extractor --bedGraph --gzip test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.bam
```

For filtering and visualising CpG level results, please see [Read-level methylation consistency analysis](#read-level-methylation-consistency).


## DETAILED DATA PROCESSING GUIDE

### Step I: UMI handling

At its 5’ end, Read 2 carries 8 bp of randomised nucleotides that serve as unique molecular identifiers (UMI) for the amplification reaction. The UMI sequence needs to be transferred from the start of Read 2 to the readID of both reads to allow UMI-aware deduplication later, a step that can be accomplished using the Trim Galore with the option --implicon (for more information type trim_galore --help). In this step, the UMI of Read 2 is added to the readID of both reads as the last element separated by a “:”, e.g.:

@HWI-D00436:407:CCAETANXX:1:1101:4105:1905 1:N:0: CGATGTTT:**CAATTTTG**

To run this specialised UMI-transfer trimming on all files of a MiSeq run you can run this command:

```
trim_galore --paired --implicon *fastq.gz
```


The FastQC per base sequence content plot would look something like this:

 	 
**Raw FastQ file:**

<img title="Read 2 - Untrimmed" style="float:right;margin:20px 20 20 20px" id="R2_untrimmed" src="Docs/Images/R2_untrimmed.png" height="300" >

**UMI trimmed file:**

<img title="Read 2 - UMI-trimmed" style="float:right;margin:20px 20 20 20px" id="R2_trimmed" src="Docs/Images/R2_UMI_trimmed.png" height="300" >


As an example, we are using the following set of test files to demonstrate subsequent steps that need to be taken:
 
**Input files:**
```
test_R1.fastq.gz
test_R2.fastq.gz
```

**Output files:**
```
test_8bp_UMI_R1.fastq.gz
test_8bp_UMI_R2.fastq.gz
```

### Step II: Adapter-/quality trimming

Following UMI-handling, UMI-treated IMPLICON reads require adapter and quality trimming. A standard Trim Galore run should identify and remove read-through adapter contamination as well as poor quality base calls, like so:

```
trim_galore --paired *UMI*fastq.gz
```

**Output files:**
```
test_8bp_UMI_R1_val_1.fq.gz
test_8bp_UMI_R2_val_2.fq.gz
```

### Step III: Genome alignments

Alignments to the mouse or human genome can then be obtained with a standard Bismark paired-end run, e.g.:

```
bismark --genome /Genomes/Mouse/GRCm38/ -1 test_8bp_UMI_R1_val_1.fq.gz -2 test_8bp_UMI_R2_val_2.fq.gz
```

#### Allele-specific alignments

> **Note for allele-specific mouse hybrid data**, 
the alignments need to be carried out against a genome that has SNPs between the parental strains masked out with Ns (N-masking). To prepare such a genome, please follow the instructions over at the [SNPsplit project](https://github.com/FelixKrueger/SNPsplit) page. The alignments as well as the output file will behave and look exactly the same, but an additional allele-sorting step needs to be carried out for allele-specific data ([see Step V below](step-v-allele-specific-sorting).

> Just briefly, if we assume a hybrid strain of Black6 (C57BL/6) and Castaneus (CAST_EiJ), the genome can be prepared using a command like this:

**Hybrid genome preparation**
```
SNPsplit_genome_preparation -vcf mgp.v5.merged.snps_all.dbSNP142.vcf.gzSNP142.vcf.gz  --strain CAST_EiJ --reference Genomes/Mouse/GRCm38/
```

**Hybrid genome indexing**

> The N-masked genome then requires indexing (as a one-off process), with a command like this:

```
bismark_genome_preparation --verbose /Genomes/Mouse/CAST_EiJ_N-masked/
```

**Allele-specific alignments**
```
bismark --genome /Genomes/Mouse/CAST_EiJ_N-masked/ -1 test_8bp_UMI_R1_val_1.fq.gz -2 test_8bp_UMI_R2_val_2.fq.gz
```


**Relevant output files:**
```
test_8bp_UMI_R1_val_1_bismark_bt2_pe.bam
```

The output BAM file is then ready for a UMI-aware deduplication step. 

### Step IV: UMI-aware deduplication

In this step, paired-end read alignments are deduplicated based on:

•	chromosome

•	start position

•	end position

•	alignment orientation

•	UMI from the read header (see [Step I](#step-i-umi-handling))

```
deduplicate_bismark --barcode test_8bp_UMI_R1_val_1_bismark_bt2_pe.bam
```

**Relevant output files:**
```
test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.bam
```

### Step V: Allele-specific sorting

> This step is optional, and only needs to be carried out for alignments against N-masked hybrid genomes processed with the     [SNPsplit package](https://github.com/FelixKrueger/SNPsplit). The command below needs the SNP annotation file for the mouse strains in question (this file is generated by the SNPsplit genome preparation):



```
SNPsplit --snp_file all_SNPs_CAST_EiJ_GRCm38.txt.gz test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.bam
```

> **Relevant output files**
```
test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.genome1.bam (C57BL/6)
test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.genome2.bam (CAST_EiJ)
```


### Step VI: Methylation extraction

The methylation extraction process can be carried out with a default parameters:

```
bismark_methylation_extractor --bedGraph --gzip test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.bam
```

> For allele-specific alignments, the files `...genome1.bam` and `...genome2.bam` should be used instead. 

**Relevant output files**

**General methylation analysis (coverage file):**
```
test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov.gz
```

**CpG context files for bisulfite consistency analysis:**
```
CpG_OT_test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.txt.gz (top strand)
CpG_OB_test_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.txt.gz (bottom strand)
```

## READ-LEVEL METHYLATION CONSISTENCY

### filtering CpG context files

This filtering step checks all CpG context result files in the current working directory for information about annotated cytosine positions of imprinted loci. The imprinted CpG annotation file (e.g. [Imprinted CpG positions Mouse (GRCm38)](CpG_imprinted_positions_mouse.txt) or [Imprinted CpG positions Human (GRCh38)](CpG_imprinted_positions_human.txt)) is provided as command line argument to the script below, and needs to contain all relevant CpG positions as a single line in this format (tab-delimited) (`Start` and `End` are the same position (= a single cytosine)):

```
Probe	Chromosome	Start	End	Feature
Chr2:152686786-152686786	2	152686786	152686786	H13
Chr2:152686810-152686810	2	152686810	152686810	H13
...
Chr2:174295708-174295708	2	174295708	174295708	Gnas
Chr2:174295728-174295728	2	174295728	174295728	Gnas
...
Chr3:34649370-34649370	3	34649370	34649370	Sox2
Chr3:34649383-34649383	3	34649383	34649383	Sox2
...
```

If a read pair is found to overlap a known imprinted locus, all CpG positions of the entire read pair are written to a file called: `methylation_state_consistency.txt`. All other reads are skipped. 

In contrast to standard methylation result files, such as the Bismark coverage file, each line of the methylation consistency file contains the CpG methylation data of an entire read-pair, and thus preserves the *methylation state on a per-read level*. The **methylation consistency file** looks like this:

```
readID	sample	implicon	1	2	3	4	5	6	7	8	9	10	11	12	13	14	15	16	17	18	19	20	21	22
	23	24	25	26	27	28	29	30	31
2	test	Klf4	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0
3	test	Klf4	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0
4	test	Klf4	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0
5	test	Sox2	NA	1	1	1	1	0	1	1 	1	1	1	1	1	1
...
```

**readID**: incremental number reads processed. **sample**: extracted from the filename via regex (might have to be adapted to different types of filenames). **implicon**: name of implicon in question (as provided by the annotation file above, last column `Feature`). **1..31**: methylation call for subsequent positions of cytosines (in CpG context) within that implicon. *0* means unmethylated, *1* means methylated. Annotated positions without a methylation call recieve '*NA*'.

The filtering scripts in the following section require Python 3 (3.6 and above).

**Command for mouse data (non allele-specific):**

```
./filter_coordinates_mouse_not_allele_specific.py CpG_imprinted_positions_mouse.txt
```

**Command for mouse data (allele-specific):**

```
./filter_coordinates_mouse_allele_specific.py CpG_imprinted_positions_mouse.txt 
```

As outlined above, allele-specific processing requires the data to have been processed with `SNPsplit`. In addition to the output for non-allelic processing above, using the script [filter_coordinates_mouse_allele_specific.py](filter_coordinates_mouse_allele_specific.py) also requires the filenames to contain `genome1` (the reference) or `genome2` (the alternative strain) which is extracted and added to the methylation consistency file as 'allele', like so:

```
readID	sample	allele	implicon	1	2	3	4	5	6	7	8	9	10	11	12	13	14	15	16	17	18	19	20	21
	22	23	24	25	26	27	28	29	30	31
1	2C_B6_miPSC	B6	Gnas	0	0	0	0	0	0	0	0	0	0
2	2C_B6_miPSC	B6	Commd1	NA	0	0	0	0	0	0	0	0	0	0	0	0
3	2C_B6_miPSC	B6	Commd1	NA	0	0	0	0	0	0	0	0	0	0	0	0
4	2C_B6_miPSC	B6	Gnas	0	1	1	1	1	0	1	1	1	1
5	2C_B6_miPSC	B6	Dlk1	NA	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	0	1	0NA	NA

```
Whereby **allele** is one of 'B6' (genome1), or 'CAST' (genome2).

**Command for human data:**

```
./filter_coordinates_human_amplicons.py CpG_imprinted_positions_human.txt
```


**Relevant output files**
```
methylation_state_consistency.txt
```

### Plotting read-level methylation consistency

In this last section, you really only need to fire up an R-Studio session and open one of the provided `Rmd` scripts, e.g. [`visualise_implicon_methylation_consistency_mouse_allele_specific.Rmd`](visualise_implicon_methylation_consistency_mouse_allele_specific.Rmd). 

Simply add the `methylation_state_consistency.txt` file to the same folder as the `Rmd` script and potentially change the filename in this line of the script:

```
read_tsv("methylation_state_consistency.txt") -> input
```

The visualisation Rmd script processes the methylation consistency file, which may contain one or several samples (and alleles for allele-specific data), etracts **Sample name** as well as **Implicon name**, and applies the following procedures:

- filters the comprehensive data table for a single Implicon
- filters for a single Sample
- transform the data to long format
- group the data by annotated CpG position, and exclude positions that were not covered at all (NA call)

- extract up to 5000 reads, 
- sort reads from highly to lowly methylated, and finally
- plot reads per gene, per sample

Either running or knitting this file to HTML format should produce various output plots which look like this:

<img title="Methylation consistency per read-pair" style="float:right;margin:20px 20 20 20px" id="methylation_consistency" src="Docs/Images/consistency_example.png" height="300" >

For any questions about this data processing guide, please get [in touch](mailto:felix.krueger@babraham.ac.uk).


[<img title="Babraham Bioinformatics" style="float:right;margin:600px 20 20 20px" id="Babraham Bioinformatics" src="Docs/Images/logo.png" height="88" >](http://www.bioinformatics.babraham.ac.uk/index.html)


CHANGELOG.md
28 March 2021

- Fixed an issue where all implicon but the very first one were shifted by 1. This applies to human and mouse, as well as the allele and non-allele-specific scripts. Thanks to Rexxi Prasaya for brining this to my attention.

- Explicitly defined the `col_character` classes for the visualisation scripts as `character` or `double`, as some columns were incorrectly auto-detected as `character` (and consequently prevented the plotting)


filter_coordinates_human_amplicons.py
#!/usr/bin/env python
import sys, gzip, os
import getopt
from glob import glob
from time import sleep
import timeit
import re


'''
This script was last changed on March 28, 2021 
'''

cpg_positions = {} # storing all relevant CpG positions for the amplicon experiment
genes = {}         # the total number of possible CpG positions per gene

# moving this out to be global variable so that reads IDs are unique
reads_processed = 0

def main():
	
	print (f"Python version: {sys.version}.")
	print (f"Reading in amplicons from file supplied as '{sys.argv[1]}'")	
	
	# get imprinted CpG positions
	read_annotation()
	 
	bismark_files = glob("CpG_*.txt.gz")
	print (f"Analysing the following Bismark CpG files:\n{bismark_files}")

	outfile = 'methylation_state_consistency.txt'
	outfh  = open (outfile,"w")

	repeat_vals = [f"{i}" for i in range(1,34)]

	# Creating a header line with 31 positional values so that we can read it into R more easily
	outfh.write ("\t".join(["readID","sample","implicon"])+ "\t" + "\t".join(repeat_vals) + "\n")

	for cpg_file in bismark_files:
		read_bismark_cpg_file(cpg_file,outfh)

	outfh.close()
	print (f"All done. Final number of reads processed: {reads_processed}\n")

def read_bismark_cpg_file(cfile,outfh):
	
	print (f"Reading file: >{cfile}<")
	
	global reads_processed

	with gzip.open(cfile) as cf:
				
		read = {}   # dictionary storing one full read and its covered positions
		old_readID = ''
		
		for line in cf:
			
			# discarding optional header
			if line.decode().startswith("Bismark"): 
				# print (f"First line: >>{line.decode().strip()}<<. Skipping...")
				continue

			# elegant solution using list comprehension. splitting to full list and then accessing afterwards
			# seems to be faster though (%%timeit). Should not be relevant here
			
			readID,state,chrom,pos = [ line.decode().strip().split(sep="\t")[i] for i in [0,1,2,3]]
			pos = int(pos)
			# print(f"ID: {readID}\tState: {state}\tChromosome: {chrom}\tPos: {pos}")
			
			if chrom in cpg_positions:
				# print (f"Found chromosome! ID: {readID}\tPos: {pos}")
				# The readID contains / characters, so let's rather rename it
				# print (readID.replace("/","_"))
				readID = readID.replace("/","_")
				
				if pos in cpg_positions[chrom]:	
					# print (f"Found it! ID: {readID}\tPos: {pos}\tGene: {cpg_positions[chrom][pos]}")
					
					if old_readID == '': # first readID
					    # print (f"Setting old_readID to {readID}")
						old_readID = readID
						read['ID'] = readID
						read['gene'] = cpg_positions[chrom][pos]
						read['filename']  = cfile
						read['positions'] = {}

					if old_readID == readID: # still the same read. Appending this position
						read['positions'][pos] = state
					else:
						# print (f"Found new read.\nOld read ID: {old_readID}\nnew read ID: {readID}\nProcessing old read now.")
						reads_processed += 1
						process_read(read,outfh,reads_processed) # process entire Read to generate graphable output
					
						# print (f"Setting up new read")
						# Resetting read dictionary
						read.clear()
						# print ("clearing read dictionary")
						# sleep(1)
						# Setting up new read
						old_readID = readID
						read['ID'] = readID
						read['gene'] = cpg_positions[chrom][pos]
						read['filename']  = cfile
						read['positions'] = {}
						read['positions'][pos] = state
					

				else:
					continue # position is not of interest as position doesn't match known positions
			else:
				continue # position is not of interest as chromosome doesn't match known positions
		
		reads_processed += 1
		if read:
			process_read(read,outfh,reads_processed) # process entire Read to generate graphable output
		
	
	
	print (f"Finished processing file {cfile}. Amplicon reads processed in total: {reads_processed}\n###############\n")
	sleep(1)


def process_read(read, outfh, reads_processed):
	
	# print (f"Got following dict: {read}")
		
	# extracting useful parts from filename
	# Example name: CpG_OT_lane1_m6_lung_ACAGTGGT_L001_22842_1_R1.UMI_val_1_GRCm38_bismark_bt2_pe.deduplicated.txt.g
	pattern = 'CpG_.+_lane\d+_(.*)_[TACG]{8}_L00.*'
	filename = read['filename']
	p = re.compile(pattern)
	# print (filename)
	m = p.findall(filename)
	sample = m[0]
	# print (sample)
	

	output = []
	output.append(reads_processed) # using a running read count instead of
	# output.append(read['ID'])    # this rather long readID to save space from the output file
	output.append(sample)
	# output.append(allele)
	output.append(read['gene']) 
	# print (f"{read['ID']}\t{sample}\t{read['gene']}\t{genes[read['gene']]}\t{len(read['positions'])}",end="\t")
	
	methylation_states = []

	for implicon_pos in genes[read['gene']]:
		# print (f"{implicon_pos}")

		if implicon_pos in read['positions'].keys():
			# print (f"{implicon_pos} and {positions[implicon_pos]}" )
			if read['positions'][implicon_pos] == '-':
				methylation_states.append(0)
			elif read['positions'][implicon_pos] == '+':
				methylation_states.append(1)
			else:
				sys.exit("Failed to get a sensible methylation state for this position")
		else:
			methylation_states.append("NA")
	
	# print (f"Methylation states: {methylation_states}")
	# joining the output
	output += methylation_states
	#  print ('\t'.join(map(str,output)))
	outfh.write('\t'.join(map(str,output)) + "\n")
	

def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)	

def read_annotation():
	
	'''
	First we need to read in the positions of CpG within the amplicons
	'''
	
	with open(sys.argv[1]) as f:

		count = 0
		
		for line in f:
			if count == 0: # 
				# eprint (f"Skipping header line >{line.strip()}<")
				# sleep(1)
				count += 1
				continue

			count += 1
			# print (f"line {count}; content >{line.strip()}<")
			# sleep(1)

			name, chrom, pos, irr, gene  = line.strip().split("\t") # irr is irrelevant
			# print (f"\t\tGene: {gene}\tChromosome: {chrom}\tPosition: {pos}")
			pos = int(pos)
			if chrom in cpg_positions.keys():
				pass
				# print (f"already present: {chrom}")
			else:
				cpg_positions[chrom] = {}

			cpg_positions[chrom][pos] = gene

			if not gene in genes:
				genes[gene] = []
			
			genes[gene].append(pos)  # keeping track of the overall number of CpG achievable per amplicon

		# print ("All chromosomes:")
		# for ch, rest in cpg_positions.items():
		# 	print (ch)
		# 	for pos in cpg_positions[ch].keys():
		# 		# print (f"pos: {pos}\tgene: {gene}")
		# 		print (f"pos: {pos}\tgene: {cpg_positions[ch][pos]}")
		# 	sleep(1)
	
	print (f"Stored the following number of CpG positions per implicon:")
	for g in genes:
	 	print (f"{g}\t{genes[g]}")

	eprint ("Finished reading annotations\n")
	sleep(3)


if __name__ == "__main__":
	main()
else:
	print ("Just getting imported")


filter_coordinates_mouse_not_allele_specific.py
#!/usr/bin/env python
import sys, gzip, os
import getopt
from glob import glob
from time import sleep
import timeit
import re

cpg_positions = {} # storing all relevant CpG positions for the amplicon experiment
genes = {}         # the total number of possible CpG positions per gene

# moving this out to be global variable so that reads IDS are unique
reads_processed = 0

def main():
	
	print (f"Python version: {sys.version}.")
	print (f"Reading in amplicons from file supplied as '{sys.argv[1]}'")	
	
	# get imprinted CpG positions
	read_annotation()
	 
	# Processing all files called CpG* in the current working directory
	bismark_files = glob("CpG_*.txt.gz")
	print (f"Analysing the following Bismark CpG files:\n{bismark_files}")

	outfile = 'methylation_state_consistency.txt'
	outfh  = open (outfile,"w")

	repeat_vals = [f"{i}" for i in range(1,32)]

	# Creating a header line with 31 positional values so that we can read it into R more easily
	outfh.write ("\t".join(["readID","sample","implicon"])+ "\t" + "\t".join(repeat_vals) + "\n")

	for cpg_file in bismark_files:
		read_bismark_cpg_file(cpg_file,outfh)

	outfh.close()
	print (f"All done. Final number of reads processed: {reads_processed}\n")

def read_bismark_cpg_file(cfile,outfh):
	
	print (f"Reading file: >{cfile}<")
	
	global reads_processed

	with gzip.open(cfile) as cf:
		
		read = {}   # dictionary storing one full read and its covered positions
		old_readID = ''
		
		for line in cf:
			
			# discarding optional header
			if line.decode().startswith("Bismark"): 
				# print (f"First line: >>{line.decode().strip()}<<. Skipping...")
				continue

			readID,state,chrom,pos = [ line.decode().strip().split(sep="\t")[i] for i in [0,1,2,3]]
			pos = int(pos)
			# print(f"ID: {readID}\tState: {state}\tChromosome: {chrom}\tPos: {pos}")
			# sleep(1)
			if chrom in cpg_positions:
				
				if pos in cpg_positions[chrom]:	
					# print (f"Found it! ID: {readID}\tPos: {pos}\tGene: {cpg_positions[chrom][pos]}")
					if old_readID == '': # first readID
						# print (f"Setting old_readID to {readID}")
						old_readID = readID
						read['ID'] = readID
						read['gene'] = cpg_positions[chrom][pos]
						read['filename']  = cfile
						read['positions'] = {}

					if old_readID == readID: # still the same read. Appending this position
						read['positions'][pos] = state
					else:
						# print (f"Found new read.\nOld read ID: {old_readID}\nnew read ID: {readID}\nProcessing old read now.")
						reads_processed += 1
						process_read(read,outfh,reads_processed) # process entire Read to generate graphable output
						
						# print (f"Setting up new read")
						# Resetting read dictionary
						read.clear()
						# print ("clearing read dictionary")
						
						### Setting up new read
						old_readID = readID
						read['ID'] = readID
						read['gene'] = cpg_positions[chrom][pos]
						read['filename']  = cfile
						read['positions'] = {}
						read['positions'][pos] = state
				else:
					continue # position is not of interest as position doesn't match known positions
			else:
				continue # position is not of interest as chromosome doesn't match known positions
		
		reads_processed += 1
		if read: # only processing if there ever was an amplicon
			process_read(read,outfh,reads_processed) # process entire Read to generate graphable output
		
	print (f"Finished processing file {cfile}. Amplicon reads processed in total: {reads_processed}\n###############\n")
	sleep(1)


def process_read(read, outfh, reads_processed):
	
	# print (f"Got following dict: {read}")
	# extracting useful parts from filename
	### This might have to be adapted for different styles of filenames

	# pattern = 'CpG_O._lane\d+_[TACG]{8}_(.*)_L00.*\.(genome[12])'
	# pattern = 'CpG_.+_lane\d+_(.*)_[TACG]{8}_L00.*'	
	# example filename: CpG_OB_lane7212_TAGCTTGT_ish4_1_noDox_C_L001_8bp_UMI_R1_val_1_bismark_bt2_pe.deduplicated.txt.gz
	# pattern = 'CpG_.+_lane\d+_[TACG]{8}_(.*)_L00.*'
	
	### Rexxi
	# CpG_OB_2_S2_L001_R1_001_val_1_bismark_bt2_pe.nonCG_filtered.txt.gz
	pattern = 'CpG_.+_(.*_.*)_L00.*'
	
	filename = read['filename']
	p = re.compile(pattern)
	# print (filename)
	m = p.findall(filename)
	sample = m[0]
	# print (sample)
	
	output = []
	output.append(reads_processed) # using a running read count instead of
	# output.append(read['ID'])    # this rather long readID to save space from the output file
	output.append(sample)
	# output.append(allele)
	output.append(read['gene']) 
	# print (f"{read_positions['ID']}\t{sample}\t{allele}\t{read_positions['gene']}\t{genes[read_positions['gene']]}\t{len(read_positions['positions'])}",end="\t")
	
	methylation_states = []

	for implicon_pos in genes[read['gene']]:
		# print (f"{implicon_pos}")

		if implicon_pos in read['positions'].keys():
			# print (f"{implicon_pos} and {read['positions'][implicon_pos]}" )
			if read['positions'][implicon_pos] == '-':
				methylation_states.append(0)
			elif read['positions'][implicon_pos] == '+':
				methylation_states.append(1)
			else:
				sys.exit("Failed to get a sensible methylation state for this position")
		else:
			# print (f"{implicon_pos} not found, assigning NA" )
			methylation_states.append("NA")
	
	# print (f"Methylation states: {methylation_states}")
	# joining the output
	output += methylation_states
	# print ('\t'.join(map(str,output)))
	outfh.write('\t'.join(map(str,output)) + "\n")
		
	
def eprint(*args, **kwargs):
	print(*args, file=sys.stderr, **kwargs)	

def read_annotation():
	
	'''
	First we need to read in the positions of CpG within the amplicons
	'''
	
	with open(sys.argv[1]) as f:

		count = 0
		
		for line in f:
			if count == 0: # 
				# eprint (f"Skipping header line >{line.strip()}<")
				# sleep(1)
				count += 1
				continue

			count += 1
			# print (f"line {count}; content >{line.strip()}<")
			# sleep(1)

			name, chrom, pos, irr, gene  = line.strip().split("\t") # irr is irrelevant
			# print (f"\t\tGene: {gene}\tChromosome: {chrom}\tPosition: {pos}")
			pos = int(pos)
			if chrom in cpg_positions.keys():
				pass
				# print (f"already present: {chrom}")
			else:
				cpg_positions[chrom] = {}

			cpg_positions[chrom][pos] = gene

			if not gene in genes:
				genes[gene] = []
			
			genes[gene].append(pos)  # keeping track of the overall number of CpG achievable per amplicon

		# print ("All chromosomes:")
		# for ch, rest in cpg_positions.items():
		# 	print (ch)
		# 	for pos in cpg_positions[ch].keys():
		# 		# print (f"pos: {pos}\tgene: {gene}")
		# 		print (f"pos: {pos}\tgene: {cpg_positions[ch][pos]}")
		# 	 sleep(1)
	
	# print (f"Stored the following number of CpG positions per implicon:")
	# for g in genes:
	#  	print (f"{g}\t{genes[g]}")

	eprint ("Finished reading annotations\n")


if __name__ == "__main__":
	main()


filter_coordinates_mouse_allele_specific.py
#!/usr/bin/env python
import sys, gzip, os
import getopt
from glob import glob
from time import sleep
import timeit
import re

'''
This script was last changed on March 28, 2021 

'''

cpg_positions = {} # storing all relevant CpG positions for the amplicon experiment
genes = {}         # the total number of possible CpG positions per gene

# moving this out to be global variable so that reads IDs are unique
reads_processed = 0

def main():
	
	print (f"Python version: {sys.version}.")
	print (f"Reading in amplicons from file supplied as '{sys.argv[1]}'")	
	
	# get imprinted CpG positions
	read_annotation()
	 
	bismark_files = glob("CpG_*.txt.gz")
	print (f"Analysing the following Bismark CpG files:\n{bismark_files}")

	outfile = 'methylation_state_consistency.txt'
	outfh  = open (outfile,"w")

	repeat_vals = [f"{i}" for i in range(1,32)]

	# Creating a header line with 31 positional values so that we can read it into R more easily
	outfh.write ("\t".join(["readID","sample","allele","implicon"])+ "\t" + "\t".join(repeat_vals) + "\n")

	for cpg_file in bismark_files:
		read_bismark_cpg_file(cpg_file,outfh)

	outfh.close()
	print (f"All done. Final number of reads processed: {reads_processed}\n")

def read_bismark_cpg_file(cfile,outfh):
	
	print (f"Reading file: >{cfile}<")
	
	global reads_processed

	with gzip.open(cfile) as cf:
				
		read = {}   # dictionary storing one full read and its covered positions
		old_readID = ''
		
		for line in cf:
			# discarding optional header
			if line.decode().startswith("Bismark"): 
				# print (f"First line: >>{line.decode().strip()}<<. Skipping...")
				continue

			readID,state,chrom,pos = [ line.decode().strip().split(sep="\t")[i] for i in [0,1,2,3]]
			pos = int(pos)
			# print(f"ID: {readID}\tState: {state}\tChromosome: {chrom}\tPos: {pos}")
			# sleep(1)
			if chrom in cpg_positions:
				
				if pos in cpg_positions[chrom]:	
					# print (f"Found it! ID: {readID}\tPos: {pos}\tGene: {cpg_positions[chrom][pos]}")
					
					if old_readID == '': # first readID
					    # print (f"Setting old_readID to {readID}")
						old_readID = readID
						read['ID'] = readID
						read['gene'] = cpg_positions[chrom][pos]
						read['filename']  = cfile
						read['positions'] = {}

					if old_readID == readID: # still the same read. Appending this position
						#' read['positions'].append(f"{state}:{pos}")
						read['positions'][pos] = state
					else:
						# print (f"Found new read.\nOld read ID: {old_readID}\nnew read ID: {readID}\nProcessing old read now.")
						reads_processed += 1
						process_read(read,outfh,reads_processed) # process entire Read to generate graphable output
						
						# print (f"Setting up new read")
						# Resetting read dictionary
						read.clear()
						# print ("clearing read dictionary")
						# sleep(1)
						# Setting up new read
						old_readID = readID
						read['ID'] = readID
						read['gene'] = cpg_positions[chrom][pos]
						read['filename']  = cfile
						read['positions'] = {}
						read['positions'][pos] = state
					
				else:
					continue # position is not of interest as position doesn't match known positions
			else:
				continue # position is not of interest as chromosome doesn't match known positions
		
		reads_processed += 1
		process_read(read,outfh,reads_processed) # process entire Read to generate graphable output
		
	
	
	print (f"Finished processing file {cfile}. Amplicon reads processed in total: {reads_processed}\n###############\n")
	sleep(1)


def process_read(read, outfh, reads_processed):
	
	# print (f"Got following dict: {read_positions}")
		
	# extracting useful parts from filename
	pattern = 'CpG_O._lane\d+_[TACG]{8}_(.*)_L00.*\.(genome[12])'
	# example read: 'CpG_OB_lane6808_TCTGCTGT_9C_A8_miPSC_L001_R1.UMI_val_1_CAST_EiJ_N-masked_GRCm38_bismark_bt2_pe.deduplicated.genome1.txt.gz'

	filename = read['filename']
	p = re.compile(pattern)
	# print (filename)
	m = p.findall(filename)
	sample = m[0][0]
	genome = m[0][1]
	
	if genome == 'genome1':
		allele = 'B6'
	elif genome == 'genome2':
		allele = 'CAST'
	else:
		sys.exit("Failed to extract allele")

	output = []
	output.append(reads_processed) # using a runnin read count instead of
	# output.append(read['ID'])    # this rather long readID to save space from the output file
	output.append(sample)
	output.append(allele)
	output.append(read['gene']) 
	#print (f"{read_positions['ID']}\t{sample}\t{allele}\t{read_positions['gene']}\t{genes[read_positions['gene']]}\t{len(read_positions['positions'])}",end="\t")
	
	methylation_states = []

	for implicon_pos in genes[read['gene']]:
		# print (f"{implicon_pos}")

		if implicon_pos in read['positions'].keys():
			# print (f"{implicon_pos} and {read['positions'][implicon_pos]}" )
			if read['positions'][implicon_pos] == '-':
				methylation_states.append(0)
			elif read['positions'][implicon_pos] == '+':
				methylation_states.append(1)
			else:
				sys.exit("Failed to get a sensible methylation state for this position")
		else:
			methylation_states.append("NA")
	
	# print (f"Methylation states: {methylation_states}")
	# joining the output
	output += methylation_states
	#  print ('\t'.join(map(str,output)))
	outfh.write('\t'.join(map(str,output)) + "\n")
		
	
def eprint(*args, **kwargs):
    print(*args, file=sys.stderr, **kwargs)	

def read_annotation():
	
	'''
	First we need to read in the positions of CpG within the amplicons
	'''
	
	with open(sys.argv[1]) as f:

		count = 0
		
		for line in f:
			if count == 0: # 
				# eprint (f"Skipping header line >{line.strip()}<")
				# sleep(1)
				count += 1
				continue

			count += 1
			# print (f"line {count}; content >{line.strip()}<")
			# sleep(1)

			name, chrom, pos, irr, gene  = line.strip().split("\t") # irr is irrelevant
			# print (f"\t\tGene: {gene}\tChromosome: {chrom}\tPosition: {pos}")
			pos = int(pos)
			if chrom in cpg_positions.keys():
				pass
				# print (f"already present: {chrom}")
			else:
				cpg_positions[chrom] = {}

			cpg_positions[chrom][pos] = gene

			if not gene in genes:
				genes[gene] = []
			
			genes[gene].append(pos)  # keeping track of the overall number of CpG achievable per amplicon

		# print ("All chromosomes:")
		# for ch, rest in cpg_positions.items():
		# 	print (ch)
		# 	for pos in cpg_positions[ch].keys():
		# 		# print (f"pos: {pos}\tgene: {gene}")
		# 		print (f"pos: {pos}\tgene: {cpg_positions[ch][pos]}")
		# 	 sleep(1)
	
	# print (f"Stored the following number of CpG positions per implicon:")
	# for g in genes:
	#  	print (f"{g}\t{genes[g]}")

	eprint ("Finished reading annotations\n")



if __name__ == "__main__":
	main()
else:
	print ("Just getting imported")
